drivers/stepper/adi_tmc/adi_tmc5041_stepper_controller.c

```diff
@@ -0,0 +1,764 @@
+	if (err != 0) {
+		return -EIO;
+	}
+
+	/* Read GSTAT register values to clear any errors SPI Datagram. */
+	uint32_t gstat_value;
+
+	err = tmc5041_read(dev, TMC5041_GSTAT, &gstat_value);
+	if (err != 0) {
+		return -EIO;
+	}
+
+	LOG_DBG("Device %s initialized", dev->name);
+
+	return 0;
+}
+
+static int tmc5041_stepper_init(const struct device *dev)
+{
+	const struct tmc5041_stepper_config *stepper_config = dev->config;
+	struct tmc5041_stepper_data *data = dev->data;
+	int err;
+
+	LOG_DBG("Controller: %s, Stepper: %s", stepper_config->controller->name, dev->name);
+
+	if (stepper_config->is_sg_enabled) {
+
+		k_work_init_delayable(&data->stallguard_dwork, stallguard_work_handler);
+
+		err = tmc5041_write(stepper_config->controller,
+				    TMC5041_SWMODE(stepper_config->index), BIT(10));
+		if (err != 0) {
+			return -EIO;
+		}
+
+		LOG_DBG("Setting stall guard to %d with delay %d ms", stepper_config->sg_threshold,
+			stepper_config->sg_velocity_check_interval_ms);
+		if (!IN_RANGE(stepper_config->sg_threshold, TMC5041_SG_MIN_VALUE,
+			      TMC5041_SG_MAX_VALUE)) {
+			LOG_ERR("Stallguard threshold out of range");
+			return -EINVAL;
+		}
+
+		int32_t stall_guard_threshold = (int32_t)stepper_config->sg_threshold;
+
+		err = tmc5041_write(
+			stepper_config->controller, TMC5041_COOLCONF(stepper_config->index),
+			stall_guard_threshold << TMC5041_COOLCONF_SG2_THRESHOLD_VALUE_SHIFT);
+		if (err != 0) {
+			return -EIO;
+		}
+		err = stallguard_enable(dev, true);
+		if (err == -EAGAIN) {
+			LOG_ERR("retrying stallguard activation");
+			k_work_reschedule(&data->stallguard_dwork,
+					  K_MSEC(stepper_config->sg_velocity_check_interval_ms));
+		}
+	}
+
+#ifdef CONFIG_STEPPER_ADI_TMC_RAMP_GEN
+	err = tmc5041_stepper_set_ramp(dev, &stepper_config->default_ramp_config);
+	if (err != 0) {
+		return -EIO;
+	}
+#endif
+
+#if CONFIG_STEPPER_ADI_TMC5041_RAMPSTAT_POLL
+	k_work_init_delayable(&data->rampstat_callback_dwork, rampstat_work_handler);
+	k_work_reschedule(&data->rampstat_callback_dwork,
+			  K_MSEC(CONFIG_STEPPER_ADI_TMC5041_RAMPSTAT_POLL_INTERVAL_IN_MSEC));
+#endif
+	err = tmc5041_stepper_set_micro_step_res(dev, stepper_config->default_micro_step_res);
+	if (err != 0) {
+		return -EIO;
+	}
+	return 0;
+}
+
+#define TMC5041_SHAFT_CONFIG(child)								\
+	(DT_PROP(child, invert_direction) << TMC5041_GCONF_SHAFT_SHIFT(DT_REG_ADDR(child)))|
+
+#define TMC5041_STEPPER_CONFIG_DEFINE(child)							\
+	COND_CODE_1(DT_PROP_EXISTS(child, stallguard_threshold_velocity),			\
+	BUILD_ASSERT(DT_PROP(child, stallguard_threshold_velocity),				\
+		     "stallguard threshold velocity must be a positive value"), ());		\
+	IF_ENABLED(CONFIG_STEPPER_ADI_TMC_RAMP_GEN, (CHECK_RAMP_DT_DATA(child)));		\
+	static const struct tmc5041_stepper_config tmc5041_stepper_config_##child = {		\
+		.controller = DEVICE_DT_GET(DT_PARENT(child)),					\
+		.default_micro_step_res = DT_PROP_OR(child, micro_step_res,			\
+						     STEPPER_MICRO_STEP_256),			\
```

Comment:
I would lay on the side of putting the default in the devicetree rather than using `DT_PROP_OR`, better location for defaults than the driver code, same for the other props, makes this macro a bit more compact if anything

Reply:
I was of the understanding that the defaults in the device tree are the register values after power-on. If this is allowed, i would be more than glad to put the defaults in the bindings.

Reply:
Have added defaults in bindings